src/bun.js/node/win_watcher.zig

```diff
@@ -165,182 +57,192 @@ pub const PathWatcherManager = struct {
+        for (this.watchers.keys()) |path| {
+            bun.default_allocator.free(path);
         }
 
-        this.file_paths.deinit();
-
-        this.watchers.deinitWithAllocator(bun.default_allocator);
-
+        this.watchers.deinit(bun.default_allocator);
         this.destroy();
     }
 };
 
+const onPathUpdateFn = JSC.Node.FSWatcher.onPathUpdate;
+const onUpdateEndFn = JSC.Node.FSWatcher.onUpdateEnd;
+
 pub const PathWatcher = struct {
     handle: uv.uv_fs_event_t,
-    ctx: ?*anyopaque,
-    recursive: bool,
-    callback: Callback,
-    flushCallback: UpdateEndCallback,
     manager: ?*PathWatcherManager,
-    path: PathWatcherManager.PathInfo,
-    last_change_event: ChangeEvent = .{},
-    closed: bool = false,
-    needs_flush: bool = false,
-
+    emit_in_progress: bool = false,
+    handlers: std.AutoArrayHashMapUnmanaged(*anyopaque, ChangeEvent) = .{},
     pub usingnamespace bun.New(PathWatcher);
 
-    const log = Output.scoped(.PathWatcher, false);
+    const log = Output.scoped(.@"fs.watch", false);
 
     pub const ChangeEvent = struct {
         hash: Watcher.HashType = 0,
         event_type: EventType = .change,
-        time_stamp: i64 = 0,
-    };
+        timestamp: u64 = 0,
+
+        pub fn emit(this: *ChangeEvent, hash: Watcher.HashType, timestamp: u64, event_type: EventType) bool {
+            const time_diff = timestamp -| this.timestamp;
+            // skip consecutive duplicates
+            if ((this.timestamp == 0 or time_diff > 1) or this.event_type != event_type and this.hash != hash) {
+                this.timestamp = timestamp;
+                this.event_type = event_type;
+                this.hash = hash;
 
-    pub const EventType = enum {
-        rename,
-        change,
-        @"error",
+                return true;
+            }
+            return false;
+        }
     };
+
+    pub const EventType = JSC.Node.FSWatcher.EventType;
     const Callback = *const fn (ctx: ?*anyopaque, path: string, is_file: bool, event_type: EventType) void;
     const UpdateEndCallback = *const fn (ctx: ?*anyopaque) void;
 
-    fn uvEventCallback(event: *uv.uv_fs_event_t, filename: [*c]const u8, events: c_int, status: c_int) callconv(.C) void {
-        if (event.data == null) return;
-        const this = bun.cast(*PathWatcher, event.data);
-
-        const manager = this.manager orelse return;
+    fn uvEventCallback(event: *uv.uv_fs_event_t, filename: ?[*:0]const u8, events: c_int, status: c_int) callconv(.C) void {
+        if (event.data == null) {
+            Output.debugWarn("uvEventCallback called with null data", .{});
+            return;
+        }
+        const this: *PathWatcher = @alignCast(@fieldParentPtr(PathWatcher, "handle", event));
+        if (comptime bun.Environment.isDebug) {
+            std.debug.assert(event.data == @as(?*anyopaque, @ptrCast(this)));
+        }
 
-        const timestamp = std.time.milliTimestamp();
+        const timestamp = event.loop.time;
```

Comment:
this time is only update when the tick starts or when we set some timer like `setTimeout`, I think is fine but if someone is doing block things this time can drift

Reply:
yeah i think thats fine